lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3893,11 +3900,9 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     assert(originalValue->getFunction() == &getOriginal());
     LLVM_DEBUG(getADDebugStream() << "Adding adjoint for " << originalValue);
 #ifndef NDEBUG
```

Comment:
Please remove `#ifndef`/`#endif` if every debug expression is in the parentheses of an `assert`.

Reply:
I see, that makes sense.

I found two other usages of `#ifndef NDEBUG`. This usage doesn't seem quite right to me:
```
  void visit(SILInstruction *inst) {
    if (errorOccurred)
      return;

    LLVM_DEBUG(getADDebugStream() << "AdjointEmitter visited:\n[ORIG]"
               << *inst);
#ifndef NDEBUG
    auto beforeInsertion = std::prev(builder.getInsertionPoint());
#endif
    SILInstructionVisitor::visit(inst);
    LLVM_DEBUG({
      auto &s = llvm::dbgs() << "[ADJ] Emitted:\n";
      auto afterInsertion = builder.getInsertionPoint();
      for (auto it = ++beforeInsertion; it != afterInsertion; ++it)
        s << *it;
    });
  }
```

Would this be more correct? (double up `NDEBUG` and `LLVM_DEBUG`)
```diff
   void visit(SILInstruction *inst) {
     if (errorOccurred)
       return;

+#ifndef NDEBUG
     LLVM_DEBUG(getADDebugStream() << "AdjointEmitter visited:\n[ORIG]"
                << *inst);
-#ifndef NDEBUG
     auto beforeInsertion = std::prev(builder.getInsertionPoint());
 #endif
     SILInstructionVisitor::visit(inst);
+#ifndef NDEBUG
     LLVM_DEBUG({
       auto &s = llvm::dbgs() << "[ADJ] Emitted:\n";
       auto afterInsertion = builder.getInsertionPoint();
       for (auto it = ++beforeInsertion; it != afterInsertion; ++it)
         s << *it;
     });
+#endif
   }
```

Reply:
No. `#ifndef NDEBUG` make code that's in the parentheses of macros that are to be compiled away compiled away. In order words, `LLVM_DEBUG` and `assert` are compiled away so you don't need `#ifndef` around them.

Reply:
> No. #ifndef NDEBUG make code that's in the parentheses of macros that are to be compiled away compiled away.

I think I understand now, thanks!